drivers: regulator: add support for RaspberryPi Pico regulator.

soburi opened the pull request:
add support for rpi_pico regulator.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>


---

Since the RaspberryPi Pico does not expose the output wiring of the Core Supply Regulator, I scraped the back of the board as shown in the picture, connected it to the ADC, and run the 'tests/drivers/regulator/voltage/' test.
![unnamed](https://user-images.githubusercontent.com/458281/216772636-b1801e2a-2496-4a98-b195-7e3ffbdf1a3c.jpg)

In RaspberryPi Pico, the Core voltage is supplied from the Core Supply Regulator, and if you disable the Core Supply Regulator or set the output to Hi-Z, the MCU will stop.
To fully test the behavior would have required designing a special board to supply the core voltage externally. So which is not tested.
From this reason, made the default settings prevent running it.

soburi committed:
drivers: regulator: add support for RaspberryPi Pico regulator.

Add support for rpi_pico regulator.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
tests: drivers: regulator: voltage: add an overlay for rpi_pico

Add overlay for RaspberryPi Pico's core-supply regulator test.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

danieldegrasse commented on drivers/regulator/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config REGULATOR_RPI_PICO
```
The description and help string of this Kconfig need to be updated

soburi replied:
Thank you for reviewing.
I updated the description.

danieldegrasse commented on drivers/regulator/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,20 @@
+if REGULATOR_RPI_PICO
+
+config REGULATOR_RPI_PICO_INIT_PRIORITY
+	int "NXP RPI_PICO regulator driver init priority"
```
Description and help string need to be updated, unless the RPI is using an NXP regulator (from the datasheet, this seems to be an internal regulator)

soburi replied:
Oops, this is embarrassing...
Fixed it.

danieldegrasse commented on drivers/regulator/regulator_rpi_pico.c:
```diff
@@ -0,0 +1,191 @@
+ * use from 850000 microvolt like Pico-SDK.
+ */
+static const struct linear_range core_ranges[] = {
+	LINEAR_RANGE_INIT(850000u, 50000u, 0u, 9u),
```
I've not familiar with this chip, but it seems to me that lower voltages could be useful if the chip was underclocked or entering a low power mode. Is there no situation where the 800000uV output would be useful?

soburi replied:
I noticed these values are usable when externally supplying the core voltage. (RP2040 usually drives by voltage from the built-in regulator, but can configure not use the regulator.)
Thus, I changed to make all values available.
However, this does not pass the test with this board.
So I made it possible to set upper and lower test voltage limits in the test so that we could exclude unstable settings from the test.

danieldegrasse replied:
I think this is a good solution, thank you for enabling this feature

danieldegrasse commented on include/zephyr/dt-bindings/regulator/rpi_pico.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_RPI_PICO_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_REGULATOR_RPI_PICO_H_
+
+#define REGULATOR_RPI_PICO_MODE_HI_Z 0x2
```
Minor nit- I would like to see a mode defined for "HI Z disabled", something like `REGULATOR_RPI_PICO_MODE_RUN`. Thi way, the `allowed-modes` and `initial-mode` properties don't have to simply be set to 0, and a user looking to change the regulator mode will have an easier time understanding that their options are RUN mode or HI Z mode

soburi replied:
Added `REGULATOR_RPI_PICO_MODE_NORMAL`.
(I think the terminology of the `RUN` cause a complex situation in case if not enable `regulator-boot-on` property. So I chose `NORMAL` for it.)

soburi committed:
tests: drivers: regulator: voltage: add min/max voltage config

Add min-microvolt/max-microvolt to set the range for voltages for testing.
It filter out ranges that cannot test correctly in the configuration.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

danieldegrasse commented on drivers/regulator/regulator_rpi_pico.c:
```diff
@@ -0,0 +1,186 @@
+
+	return linear_range_group_get_value(
+		config->ranges, config->num_ranges,
+		((*vreg & VREG_AND_CHIP_RESET_VREG_VSEL_BITS) >> VREG_AND_CHIP_RESET_VREG_VSEL_LSB),
```
I don't have a board to test on, but won't you need to subtract `VREG_VSEL_IDX_OFFSET` here to get the correct voltage? As an aside, the linear range API also supports multiple values mapping to the same voltage (see here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/regulator/regulator_pca9420.c#L127), which might be the best way to handle this regulator

soburi replied:
Thanks for suggestion.
I rewrite with correctly linear range API usage, and removed `VREG_VSEL_IDX_OFFSET`.

gmarull commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -184,6 +185,22 @@
+			raspberrypi,brown-out-detection;
+			raspberrypi,brown-out-threshold = <860000>;
+			vreg_core: CORE {
+				regulator-always-on;
+				regulator-boot-on;
+				regulator-initial-mode = <REGULATOR_RPI_PICO_MODE_NORMAL>;
+				regulator-allowed-modes = <REGULATOR_RPI_PICO_MODE_NORMAL>;
+				regulator-min-microvolt = <800000>;
+				regulator-max-microvolt = <1300000>;
```
note: these flags are used to limit regulator operating range, not to describe hardware capabilities. So in general, they should be put at board level. Since this is put at the SoC DTS files, it looks like for example the always-on/boot-on is something that can't be configured, so driver doesn't need to implement enable/disable.

soburi replied:
I moved these properties into boards/.
And it left properties that need to keep the regulator from turning off only.

gmarull commented on drivers/regulator/regulator_rpi_pico.c:
```diff
@@ -0,0 +1,185 @@
+	.get_mode = regulator_rpi_pico_get_mode,
+};
+
+static const struct regulator_driver_api parent_api = {
+	/* place holder */
+};
```
note: parent device is not required, just drop it if not needed.

soburi replied:
Delete the parent device and make a flattened structure.

gmarull commented on drivers/regulator/regulator_rpi_pico.c:
```diff
@@ -0,0 +1,185 @@
+
+struct regulator_rpi_pico_config {
+	struct regulator_common_config common;
+	const struct linear_range *ranges;
+	const size_t num_ranges;
```
note: it looks like you only have one regulator (CORE), so there's no need to store ranges in the instance config struct. Just use `core_ranges` directly.

soburi replied:
Make it use `core_ranges` and `num_core_ranges` directly.

gmarull commented on drivers/regulator/regulator_rpi_pico.c:
```diff
@@ -0,0 +1,185 @@
+static int regulator_rpi_pico_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+	volatile uint32_t *vreg = &(((volatile vreg_and_chip_reset_hw_t *)config->reg)->vreg);
```
nit: just make config->reg a `vreg_and_chip_reset_hw_t *` pointer

soburi replied:
Applied it.

gmarull reviewed (changes requested):
thanks for this change, a few questions/change requests

gmarull commented on tests/drivers/regulator/voltage/dts/bindings/test-regulator-voltage.yaml:
```diff
@@ -42,3 +42,17 @@ properties:
+
+  min-microvolt:
+    type: int
+    default: 0
```
volts can take `int32_t` range (includes negatives)

soburi replied:
Change default to INT32_MIN.

gmarull commented on tests/drivers/regulator/voltage/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,35 @@
+	resources: resources {
+		compatible = "test-regulator-voltage";
+		regulators = <&vreg_core>;
+		io-channels = <&adc 0>;
```
maybe add a comment on how to wire this up (seems you need to solder a wire)

soburi replied:
I can't attach a picture here, so I described it in the comments.

soburi committed:
boards: arm: rpi_pico: add regulator definition on dts

Add a regulator node to keep it always on.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: arm: adafruit_kb20240: add regulator definition on dts

Add a regulator node to keep it always on.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: arm: sparkfun_pro_micro_rp2040: add regulator definition on dts

Add a regulator node to keep it always on.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi commented:
@danieldegrasse
Could you please re-approve this PR if there are no problems?
